apps/process_tpstream.cxx

```diff
@@ -264,69 +280,21 @@ int main(int argc, char const *argv[])
       //
       // TA Processing
       //
-      // Create the output buffer
-      std::vector<triggeralgs::TriggerActivity> ta_buffer;
-
-      // Loop over TPs
-      size_t n_tas = 0;
 
-      for( const auto& tp : tp_buffer ) {
-
-        if ( tp_filter(tp) ){
-          // if(!quiet)
-              // fmt::print("  TP filtered out!");
-          continue;
-        }
-
-        (*ta_maker)(tp, ta_buffer);
-
-        if (n_tas != ta_buffer.size()) {
-          if (!quiet) {
-            for( size_t i=n_tas; i<ta_buffer.size(); ++i){
-              const auto& ta = ta_buffer[i];
-              fmt::print("  + {} : ca_s={} ca_e={} ta_s={}, ta_e={} d_ta = {} | tp_s={} tp_s-ta_s={}\n", i, ta.channel_start, ta.channel_end, ta.time_start, ta.time_end, ta.time_end-ta.time_start, tp.time_start, (tp.time_start-ta.time_start) );
-            }
-          }
-          n_tas = ta_buffer.size();
-        }
-      }
+      const auto ta_start = std::chrono::steady_clock::now();
+      std::unique_ptr<daqdataformats::Fragment> ta_frag = ta_emulator->emulate(tp_buffer);
+      const auto ta_end = std::chrono::steady_clock::now();
 
-      // Count how many TAs were generated
-      if (!quiet)
-        fmt::print("  ta_buffer.size() = {}\n", ta_buffer.size());
-
-      // Their size
-      size_t payload_size(0);
-      for ( const auto& ta : ta_buffer ) {
-        payload_size += triggeralgs::get_overlay_nbytes(ta);
-      }
-
-      // Print the total size
-      if (!quiet)
-        fmt::print("  ta_buffer in bytes = {}\n", payload_size);
-
-      // Could be that a TA is empty; skip as it will show up in the next fragment.
-      // TC also gets skipped, but what's the use of trying to propagate a TC with no TAs?
-      if (payload_size == 0) {
-        if (!quiet)
-          fmt::print("Skipped saving an empty TA frag.");
+      if (ta_frag == nullptr) // Buffer was empty.
         continue;
-      }
-      
-      // Create the fragment buffer
-      void* payload = malloc(payload_size);
-
-      size_t payload_offset(0);
-      for ( const auto& ta : ta_buffer ) {
-        triggeralgs::write_overlay(ta, payload + payload_offset);
-        payload_offset += triggeralgs::get_overlay_nbytes(ta);
-      }
-
-      // Hand it to the fragment
-      std::unique_ptr<daqdataformats::Fragment> ta_frag = std::make_unique<daqdataformats::Fragment>(payload, payload_size);
+      num_tas += ta_emulator->get_last_output_buffer().size();
 
-      // And release it
-      free(payload);
+      // TA time calculation.
+      const uint64_t ta_diff = std::chrono::nanoseconds(ta_end - ta_start).count();
+      average_ta_time += ta_diff;
+      if (!quiet) {
+        fmt::print("\tTA Time Process: {} ns.\n", ta_diff);
+      }
```

Comment:
Not for now, but would be nice to somehow save all the latencies for later histogramming... no idea how to do it atm without using text files, root, or something even worse.